### PR TITLE
Converted exp_homun database to yaml

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: -

* **Server Mode**: both

* **Description of Pull Request**: 

  * Converts the Homunculus Experience Database into YAML.
  * Includes CSV2YAML converter.

### CODE REVIEW

File: src/map/homunculus.cpp [hunk 1]
```diff
@@ -1323,7 +1381,7 @@ void hom_reset_stats(struct homun_data *hd)
 	hom->dex = base->dex *10;
 	hom->luk = base->luk *10;
 	hom->exp = 0;
-	hd->exp_next = hexptbl[0];
+	hd->exp_next = homun_exp_db.get_nextexp(1);
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:1384]:
```suggestion
	hd->exp_next = homun_exp_db.get_nextexp(hom->level);
```

File: src/map/homunculus.cpp [hunk 2]
```diff
@@ -1613,52 +1671,9 @@ static void read_homunculus_skilldb(void) {
...
-		}
-		fclose(fp);
-		ShowStatus("Done reading '" CL_WHITE "%d" CL_RESET "' levels in '" CL_WHITE "%s" CL_RESET "'.\n", j, path);
-	}
-}
-
 void hom_reload(void){
 	read_homunculusdb();
-	read_homunculus_expdb();
+	homun_exp_db.load();
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:1676]:
```suggestion
	homun_exp_db.reload();
```

**Review by Lemongrass3110 — COMMENTED** [review]
It would be nice if we could get rid of the requirement to have "0" as dummy entry for the last level.

File: src/map/homunculus.cpp [hunk 3]
```diff
@@ -51,6 +50,66 @@ static struct s_homun_intimacy_grade intimacy_grades[] = {
...
+	uint64 exp;
+
+	if (!this->asUInt64(node, "Exp", exp))
+		return 0;
+
+	if (level == 0) {
+		this->invalidWarning(node["Level"], "The minimum level is 1.\n");
+		return 0;
+	}
+	if (level > MAX_LEVEL) {
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:76]:
```suggestion
	if (level >= MAX_LEVEL) {
```

File: src/map/homunculus.cpp [hunk 4]
```diff
@@ -51,6 +50,66 @@ static struct s_homun_intimacy_grade intimacy_grades[] = {
...
+ * @param level: Homunculus level
+ * @return Experience
+ */
+t_exp HomExpDatabase::get_nextexp(uint16 level) {
+	if (level < MAX_LEVEL) {
+		auto next_exp = this->find(level);
+		if (next_exp)
+			return next_exp->exp;
+	}
+	return 0;
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:110]:
```suggestion
	auto next_exp = this->find(level);
	if (next_exp)
		return next_exp->exp;
	else
		return 0;
```

File: src/tool/csv2yaml.cpp [hunk 5]
```diff
@@ -3451,3 +3463,35 @@ static bool mob_parse_row_chatdb(char* fields[], int columns, int current) {
...
+
+		if (line[0] == '/' && line[1] == '/') // Ignore comments
+			continue;
+
+		t_exp exp = strtoull(line, nullptr, 10);
+
+		body << YAML::BeginMap;
+		body << YAML::Key << "Level" << YAML::Value << (count+1);
+		body << YAML::Key << "Exp" << YAML::Value << exp;
+		body << YAML::EndMap;
```
Comment (Lemongrass3110) [inline src/tool/csv2yaml.cpp:3489]:
```suggestion
		if( exp > 0 ){
			body << YAML::BeginMap;
			body << YAML::Key << "Level" << YAML::Value << (count+1);
			body << YAML::Key << "Exp" << YAML::Value << exp;
			body << YAML::EndMap;
		}
```

File: db/exp_homun.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#
+###########################################################################
+# Homunculus Experience Database
+###########################################################################
+#
+# Homunculus Experience Settings
+#
+###########################################################################
+# - Level   Level required.
+#   Exp     Experience required for the next level. 0: no next level (default 0).
```
Comment (Lemongrass3110) [inline db/exp_homun.yml:26]:
```suggestion
#   Exp     Experience required for the next level.
```

File: db/import-tmpl/exp_homun.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+#
+###########################################################################
+# Homunculus Experience Database
+###########################################################################
+#
+# Homunculus Experience Settings
+#
+###########################################################################
+# - Level   Level required.
+#   Exp     Experience required for the next level. 0: no next level (default 0).
```
Comment (Lemongrass3110) [inline db/import-tmpl/exp_homun.yml:26]:
```suggestion
#   Exp     Experience required for the next level.
```

File: db/pre-re/exp_homun.yml [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+#
+###########################################################################
+# Homunculus Experience Database
+###########################################################################
+#
+# Homunculus Experience Settings
+#
+###########################################################################
+# - Level   Level required.
+#   Exp     Experience required for the next level. 0: no next level (default 0).
```
Comment (Lemongrass3110) [inline db/pre-re/exp_homun.yml:26]:
```suggestion
#   Exp     Experience required for the next level.
```

File: db/re/exp_homun.yml [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+#
+###########################################################################
+# Homunculus Experience Database
+###########################################################################
+#
+# Homunculus Experience Settings
+#
+###########################################################################
+# - Level   Level required.
+#   Exp     Experience required for the next level. 0: no next level (default 0).
```
Comment (Lemongrass3110) [inline db/re/exp_homun.yml:26]:
```suggestion
#   Exp     Experience required for the next level.
```

File: doc/yaml/db/exp_homun.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+###########################################################################
+# Homunculus Experience Database
+###########################################################################
+#
+# Homunculus Experience Settings
+#
+###########################################################################
+# - Level   Level required.
+#   Exp     Experience required for the next level. 0: no next level (default 0).
```
Comment (Lemongrass3110) [inline doc/yaml/db/exp_homun.yml:9]:
```suggestion
#   Exp     Experience required for the next level.
```

File: src/map/homunculus.cpp (hunk 3 above)
Comment (Atemo) [reply src/map/homunculus.cpp:76]: ha yes it's smarter!

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**